protostar-rust/src/lib.rs

```diff
@@ -1,84 +1,148 @@
+
+pub mod pretty_printing;
+
+#[derive(Default)]
+pub struct TestsStats {
+    passed: usize,
+    failed: usize,
+}
+
 #[derive(Deserialize, Debug)]
 pub struct ProtostarTestConfig {
     #[serde(default)]
     exit_first: bool, // TODO Not implemented!
 }
 
-fn run_result_value_to_string(run_result: RunResultValue) -> String {
-    return match run_result {
-        RunResultValue::Success(data) => format!("PASS {:?}", data),
-        RunResultValue::Panic(data) => format!("FAIL {:?}", data),
-    };
+fn result_data_to_text(data: &[Felt252]) -> String {
+    let mut readable_text = String::new();
+
+    for felt in data {
+        let felt_bytes = felt.to_bytes_be();
+        let felt_text = String::from_utf8_lossy(&felt_bytes);
+        readable_text.push_str(&*felt_text);
+    }
+
+    readable_text
+}
+
+fn get_result_str_and_update_tests_stats(
+    run_result: RunResultValue,
+    tests_stats: &mut TestsStats,
+) -> String {
+    match run_result {
+        RunResultValue::Success(result_data) => {
+            tests_stats.passed += 1;
+            result_data_to_text(&result_data)
+        }
+        RunResultValue::Panic(result_data) => {
+            tests_stats.failed += 1;
+            result_data_to_text(&result_data)
+        }
+    }
 }
 
 fn collect_tests_in_directory(input_path: &Utf8PathBuf) -> Result<Vec<Utf8PathBuf>> {
-    let mut test_directories: Vec<Utf8PathBuf> = vec![];
+    let mut test_files: Vec<Utf8PathBuf> = vec![];
 
     for entry in WalkDir::new(input_path) {
         let entry =
             entry.with_context(|| format!("Failed to read directory at path = {}", input_path))?;
         let path = entry.path();
 
         if path.is_file() && path.extension().unwrap_or_default() == "cairo" {
-            test_directories.push(
+            test_files.push(
                 Utf8Path::from_path(path)
                     .with_context(|| format!("Failed to convert path = {:?} to utf-8", path))?
                     .to_path_buf(),
             );
         }
     }
 
-    Ok(test_directories)
+    Ok(test_files)
 }
 
-pub fn run_tests(
+pub fn run_test_runner(
     input_path: Utf8PathBuf,
     linked_libraries: Option<Vec<LinkedLibrary>>,
     config: &ProtostarTestConfig,
 ) -> Result<()> {
     let test_directories = collect_tests_in_directory(&input_path)?;
-    for test in test_directories {
-        run_tests_in_file(test, linked_libraries.clone())?;
+
+    let mut tests_vector = vec![];
+
+    let builtins = vec!["GasBuiltin", "Pedersen", "RangeCheck", "bitwise", "ec_op"];
+    for test_file in test_directories {
+        let (sierra_program, test_configs) = collect_tests(
+            test_file.as_str(),
+            None,
+            linked_libraries.clone(),
+            Some(builtins.clone()),
+        )?;
+        let a = test_file.clone();
+        let x = a.strip_prefix(input_path.clone())?.to_path_buf();
+        tests_vector.push((sierra_program, test_configs, x));
```

Comment:
This could be abstracted to a function

Reply:
I disagree, it would require us to pass 2 paths around (directory path + test file path) instead of one. Change it to be more readable now, check it out

Reply:
I don't think having to pass arguments around should prevent us from defining functions for things.